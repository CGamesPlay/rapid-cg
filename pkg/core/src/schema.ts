export type ColumnAny = { name: string; type: string; primary?: boolean };
export type ColumnDate = ColumnAny & {
  type: "date";
  mode?: "createdAt" | "updatedAt";
};
export type Column = ColumnAny | ColumnDate;

export type Table = {
  name: string;
  columns: Record<string, Column>;
};

export type Database = {
  tables: Record<string, Table>;
};

function clone<T>(base: T, ...newProps: Array<any>): T {
  const ret = Object.create(Object.getPrototypeOf(base));
  Object.assign(ret, ...newProps);
  return ret;
}

class ColumnAnyBuilder {
  result: Omit<Column, "name">;

  constructor(input: Partial<Column>) {
    this.result = input as Omit<Column, "name">;
    Object.assign(this, input);
  }

  build(name: string): Column {
    return Object.assign({}, this.result, { name });
  }

  protected withProperties(input: Partial<Column>): this {
    const ret = Object.create(Object.getPrototypeOf(this));
    ret.result = {};
    Object.assign(ret.result, this.result, input);
    return ret;
  }

  primary(primary = true): this {
    return this.withProperties({ primary });
  }
}

class ColumnDateBuilder extends ColumnAnyBuilder {
  mode: undefined | "createdAt" | "updatedAt";

  createdAt(): this {
    return this.withProperties({ mode: "createdAt" });
  }

  updatedAt(): this {
    return this.withProperties({ mode: "updatedAt" });
  }
}

function uuid(): ColumnAnyBuilder {
  return new ColumnAnyBuilder({ type: "uuid" });
}

function date(): ColumnDateBuilder {
  return new ColumnDateBuilder({ type: "date" });
}

function text(): ColumnAnyBuilder {
  return new ColumnAnyBuilder({ type: "text" });
}

class TableBuilder {
  columns: Record<string, Column> = {};

  build(name: string): Table {
    return Object.assign({}, this, { name }) as any;
  }

  withTimestamps(): this {
    return clone(this, {
      columns: {
        ...this.columns,
        createdAt: s.date().createdAt(),
        updatedAt: s.date().updatedAt(),
      },
    });
  }
}

function table(columns: Record<string, ColumnAnyBuilder>): TableBuilder {
  const result = new TableBuilder();
  for (let name in columns) {
    result.columns[name] = columns[name].build(name);
  }
  return result;
}

function database(tables: Record<string, TableBuilder>): Database {
  const result: Database = { tables: {} as any };
  for (let name in tables) {
    result.tables[name] = tables[name].build(name);
  }
  return result;
}

export const s = {
  uuid,
  date,
  text,
  table,
  database,
};

export default s;
