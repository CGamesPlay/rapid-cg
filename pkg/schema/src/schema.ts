export type ColumnAny = {
  name: string;
  unique?: boolean;
  nullable?: boolean;
};
export type ColumnDate = ColumnAny & {
  type: "date";
  mode?: "createdAt" | "updatedAt";
  primary?: boolean;
  default?: Date;
};
export type ColumnInteger = ColumnAny & {
  type: "integer";
  primary?: boolean | "autoincrement";
  default?: number | bigint;
};
export type ColumnJson = ColumnAny & {
  type: "json";
  primary?: boolean;
  default?: unknown;
};
export type ColumnText = ColumnAny & {
  type: "text";
  primary?: boolean;
  default?: string;
};
export type ColumnUuid = ColumnAny & {
  type: "uuid";
  autogenerate?: boolean;
  primary?: boolean;
  default?: string;
};
export type Column =
  | ColumnDate
  | ColumnInteger
  | ColumnJson
  | ColumnText
  | ColumnUuid;

export type Table = {
  name: string;
  columns: Record<string, Column>;
};

export type Database = {
  tables: Record<string, Table>;
};

function clone<T>(base: T, ...newProps: Array<any>): T {
  const ret = Object.create(Object.getPrototypeOf(base));
  Object.assign(ret, ...newProps);
  return ret;
}

class ColumnAnyBuilder<DefaultType = never> {
  result: Omit<Column, "name">;

  constructor(input: Omit<Column, "name">) {
    this.result = input;
    Object.assign(this, input);
  }

  build(name: string): Column {
    return Object.assign({} as Column, this.result, { name });
  }

  protected withProperties(input: unknown): this {
    const ret = Object.create(Object.getPrototypeOf(this));
    ret.result = {};
    Object.assign(ret.result, this.result, input);
    return ret;
  }

  primary(primary = true): this {
    return this.withProperties({ primary });
  }

  unique(unique = true): this {
    return this.withProperties({ unique });
  }

  nullable(nullable = true): this {
    return this.withProperties({ nullable });
  }

  default(value: DefaultType): this {
    return this.withProperties({ default: value });
  }
}

class ColumnUuidBuilder extends ColumnAnyBuilder<string> {
  autogenerate(autogenerate = true): this {
    return this.withProperties({ autogenerate });
  }
}

class ColumnDateBuilder extends ColumnAnyBuilder<Date> {
  mode: undefined | "createdAt" | "updatedAt";

  createdAt(): this {
    return this.withProperties({ mode: "createdAt" });
  }

  updatedAt(): this {
    return this.withProperties({ mode: "updatedAt" });
  }
}

class ColumnIntegerBuilder extends ColumnAnyBuilder<number | bigint> {
  autoincrement(): this {
    return this.withProperties({ primary: "autoincrement" });
  }
}

class TableBuilder {
  columns: Record<string, Column> = {};

  build(name: string): Table {
    return Object.assign({}, this, { name }) as any;
  }

  withTimestamps(): this {
    return clone(this, {
      columns: {
        ...this.columns,
        createdAt: s.date().createdAt(),
        updatedAt: s.date().updatedAt(),
      },
    });
  }
}

export const s = {
  date(): ColumnDateBuilder {
    return new ColumnDateBuilder({ type: "date" });
  },

  integer(): ColumnIntegerBuilder {
    return new ColumnIntegerBuilder({ type: "integer" });
  },

  json(): ColumnAnyBuilder<unknown> {
    return new ColumnAnyBuilder({ type: "json" });
  },

  text(): ColumnAnyBuilder<string> {
    return new ColumnAnyBuilder({ type: "text" });
  },

  uuid(): ColumnUuidBuilder {
    return new ColumnUuidBuilder({ type: "uuid" });
  },

  table(columns: Record<string, ColumnAnyBuilder<never>>): TableBuilder {
    const result = new TableBuilder();
    for (let name in columns) {
      result.columns[name] = columns[name].build(name);
    }
    return result;
  },

  database(tables: Record<string, TableBuilder>): Database {
    const result: Database = { tables: {} as any };
    for (let name in tables) {
      result.tables[name] = tables[name].build(name);
    }
    return result;
  },
};

export default s;
