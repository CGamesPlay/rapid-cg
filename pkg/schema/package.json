{
  "name": "@rad/schema",
  "version": "0.5.0",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "license": "MIT",
  "dependencies": {
    "commander": "^8.3.0",
    "lodash": "^4.17.21",
    "pluralize": "^8.0.0",
    "zod": "^3.11.6"
  },
  "devDependencies": {
    "@types/lodash": "^4.14.176",
    "@types/pluralize": "^0.0.29"
  }
}
