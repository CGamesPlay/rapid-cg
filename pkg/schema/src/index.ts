import { Command } from "commander";

import { DatabaseSchema } from "./types.js";

export * from "./types.js";
export * from "./schema.js";
export * from "./utils.js";

export interface Generator {
  // Should correspond to the package name, e.g. "@rad/sqlite-generator"
  name: string;
  // Runs this generator.
  generate?(config: Config): Promise<void>;
  // Return all of the additional commands that this generator supports.
  commands?(config: Config): Command[];
}

export type Config = {
  database: DatabaseSchema;
  generators: Generator[];
};
