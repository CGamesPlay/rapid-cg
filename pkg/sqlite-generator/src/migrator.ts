import { s, DatabaseSchema, TableSchema, Column } from "@rad/schema";

function id(val: string) {
  return `"${val.replace(/"/g, '""')}"`;
}

function columnType(column: Column): string {
  switch (column.type) {
    case "date":
    case "json":
    case "text":
    case "uuid":
      return "TEXT";
    case "integer":
      return "INTEGER";
    /* istanbul ignore next */
    default:
      throw new Error(`Unsupported column type ${(column as any).type}`);
  }
}

function columnConstraints(column: Column): string {
  const constraints: string[] = [];
  if (column.primary) {
    constraints.push("PRIMARY KEY");
    if (column.primary === "autoincrement") {
      constraints.push("AUTOINCREMENT");
    }
  }
  if (column.unique) {
    constraints.push("UNIQUE");
  }
  if (!column.nullable) {
    constraints.push("NOT NULL");
  }
  return constraints.join(" ");
}

function columnDefinition(c: Column): string {
  return [id(c.name), columnType(c), columnConstraints(c)]
    .filter((x) => x)
    .join(" ");
}

function generateTableCopyMigration(
  from: TableSchema,
  to: TableSchema
): string {
  const transferTable = `transfer${to.name}`;
  let lines: string[] = [];
  lines = lines.concat([
    `BEGIN EXCLUSIVE TRANSACTION;`,
    `CREATE TABLE ${id(transferTable)} (`,
    Object.values(to.columns)
      .map((c) => `  ${columnDefinition(c)}`)
      .join(",\n"),
    `);`,
  ]);
  const sharedColumns: string[] = [];
  for (let column in from.columns) {
    if (to.columns[column]) {
      sharedColumns.push(column);
    }
  }
  lines = lines.concat([
    `INSERT INTO ${transferTable} ( ${sharedColumns.map(id).join(", ")} )`,
    `  SELECT ${sharedColumns.map(id).join(", ")}`,
    `  FROM ${id(from.name)};`,
    `DROP TABLE ${from.name};`,
    `ALTER TABLE ${transferTable} RENAME TO ${to.name};`,
    `COMMIT TRANSACTION;`,
  ]);
  return lines.join("\n");
}

function generateTableMigration(
  from: TableSchema | undefined,
  to: TableSchema | undefined
): string {
  /* istanbul ignore if */
  if (from === undefined && to === undefined) return "";
  else if (from !== undefined && to !== undefined) {
    const statements: string[] = [];
    for (let column in from.columns) {
      if (
        from.columns[column] &&
        to.columns[column] &&
        columnDefinition(from.columns[column]) ===
          columnDefinition(to.columns[column])
      ) {
        continue;
      } else if (to.columns[column]) {
        return generateTableCopyMigration(from, to);
      } else {
        statements.push(
          `ALTER TABLE ${id(from.name)} DROP COLUMN ${id(column)};`
        );
      }
    }
    for (let column in to.columns) {
      if (!(column in from.columns)) {
        statements.push(
          `ALTER TABLE ${id(from.name)} ADD COLUMN ${columnDefinition(
            to.columns[column]
          )};`
        );
      }
    }
    return statements.join("\n\n");
  } else if (to !== undefined) {
    return [
      `CREATE TABLE ${id(to.name)} (`,
      Object.values(to.columns)
        .map((c) => `  ${columnDefinition(c)}`)
        .join(",\n"),
      `);`,
    ].join("\n");
  } else if (from !== undefined) {
    return `DROP TABLE ${id(from.name)};`;
  } else {
    /* istanbul ignore next */
    return "";
  }
}

type MigrationParams = { from: DatabaseSchema; to: DatabaseSchema };

export function generateMigration({ from, to }: MigrationParams): string {
  const components: string[] = [];
  for (let table in from.tables) {
    components.push(
      generateTableMigration(from.tables[table], to.tables[table])
    );
  }
  for (let table in to.tables) {
    if (!(table in from.tables)) {
      components.push(generateTableMigration(undefined, to.tables[table]));
    }
  }
  return components.filter((c) => c).join("\n\n");
}
