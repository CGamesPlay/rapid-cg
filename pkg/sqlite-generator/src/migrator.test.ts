import { s } from "@rad/schema";

import { generateMigration } from "./migrator.js";

describe("generateMigration", () => {
  it("creates a table", () => {
    const src = generateMigration({
      from: s.database({}),
      to: s.database({
        User: s.model({
          id: s.integer().autoincrement(),
        }),
      }),
    });
    expect(src).toMatchInlineSnapshot(`
      "CREATE TABLE \\"users\\" (
        \\"id\\" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL
      );"
    `);
  });

  it("handles table aliases", () => {
    const src = generateMigration({
      from: s.database({
        Tbl: s.model({ id: s.integer().autoincrement() }).inTable("tbl"),
      }),
      to: s.database({
        MyModel: s.model({ id: s.integer().autoincrement() }).inTable("tbl"),
      }),
    });
    expect(src).toEqual("");
  });

  it("supports all column types", () => {
    const src = generateMigration({
      from: s.database({}),
      to: s.database({
        User: s.model({
          boolean: s.boolean(),
          date: s.date(),
          integer: s.integer(),
          json: s.json(),
          text: s.text(),
          uuid: s.uuid(),
        }),
      }),
    });
    expect(src).toMatchInlineSnapshot(`
      "CREATE TABLE \\"users\\" (
        \\"boolean\\" INTEGER NOT NULL,
        \\"date\\" TEXT NOT NULL,
        \\"integer\\" INTEGER NOT NULL,
        \\"json\\" TEXT NOT NULL,
        \\"text\\" TEXT NOT NULL,
        \\"uuid\\" TEXT NOT NULL
      );"
    `);
  });

  it("supports all column constraints", () => {
    const src = generateMigration({
      from: s.database({}),
      to: s.database({
        User: s.model({
          unique: s.integer().unique(),
          nullable: s.integer().nullable(),
        }),
      }),
    });
    expect(src).toMatchInlineSnapshot(`
      "CREATE TABLE \\"users\\" (
        \\"unique\\" INTEGER UNIQUE NOT NULL,
        \\"nullable\\" INTEGER
      );"
    `);
  });

  it("creates with foreign keys", () => {
    const src = generateMigration({
      from: s.database({}),
      to: s.database({
        User: s.model({
          id: s.integer().autoincrement(),
          parentId: s.integer().nullable(),
          parent: s.belongsTo("parentId", "User", "id"),
          children: s.hasMany("id", "User", "parentId"),
        }),
      }),
    });
    expect(src).toMatchInlineSnapshot(`
      "CREATE TABLE \\"users\\" (
        \\"id\\" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
        \\"parentId\\" INTEGER,
        FOREIGN KEY ( \\"parentId\\" ) REFERENCES \\"users\\" ( \\"id\\" )
      );"
    `);
  });

  it("drops a table", () => {
    const src = generateMigration({
      from: s.database({
        User: s.model({ col: s.text() }),
      }),
      to: s.database({}),
    });
    expect(src).toMatchInlineSnapshot(`"DROP TABLE \\"users\\";"`);
  });

  it("adds a column", () => {
    const src = generateMigration({
      from: s.database({
        User: s.model({ col: s.text() }),
      }),
      to: s.database({
        User: s.model({ col: s.text(), extra: s.text() }),
      }),
    });
    expect(src).toMatchInlineSnapshot(
      `"ALTER TABLE \\"users\\" ADD COLUMN \\"extra\\" TEXT NOT NULL;"`
    );
  });

  it("drops a column", () => {
    const src = generateMigration({
      from: s.database({
        User: s.model({ col: s.text(), extra: s.text() }),
      }),
      to: s.database({
        User: s.model({ col: s.text() }),
      }),
    });
    expect(src).toMatchInlineSnapshot(
      `"ALTER TABLE \\"users\\" DROP COLUMN \\"extra\\";"`
    );
  });

  it("changes a column", () => {
    const src = generateMigration({
      from: s.database({
        User: s.model({ col: s.text(), extra: s.text() }),
      }),
      to: s.database({
        User: s.model({ col: s.text(), extra: s.text().nullable() }),
      }),
    });
    expect(src).toMatchInlineSnapshot(`
      "BEGIN EXCLUSIVE TRANSACTION;
      CREATE TABLE \\"transferusers\\" (
        \\"col\\" TEXT NOT NULL,
        \\"extra\\" TEXT
      );
      INSERT INTO transferusers ( \\"col\\", \\"extra\\" )
        SELECT \\"col\\", \\"extra\\"
        FROM \\"users\\";
      DROP TABLE users;
      ALTER TABLE transferusers RENAME TO users;
      COMMIT TRANSACTION;"
    `);
  });

  it("alters a foreign key", () => {
    const src = generateMigration({
      from: s.database({
        User: s.model({
          id: s.integer().autoincrement(),
          parentId: s.integer().nullable(),
        }),
      }),
      to: s.database({
        User: s.model({
          id: s.integer().autoincrement(),
          parentId: s.integer().nullable(),
          parent: s.belongsTo("parentId", "User", "id"),
          children: s.hasMany("id", "User", "parentId"),
        }),
      }),
    });
    expect(src).toMatchInlineSnapshot(`
      "BEGIN EXCLUSIVE TRANSACTION;
      CREATE TABLE \\"transferusers\\" (
        \\"id\\" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
        \\"parentId\\" INTEGER,
        FOREIGN KEY ( \\"parentId\\" ) REFERENCES \\"users\\" ( \\"id\\" )
      );
      INSERT INTO transferusers ( \\"id\\", \\"parentId\\" )
        SELECT \\"id\\", \\"parentId\\"
        FROM \\"users\\";
      DROP TABLE users;
      ALTER TABLE transferusers RENAME TO users;
      COMMIT TRANSACTION;"
    `);
  });

  it("drops a foreign key", () => {
    const src = generateMigration({
      from: s.database({
        User: s.model({
          id: s.integer().autoincrement(),
          parentId: s.integer().nullable(),
          parent: s.belongsTo("parentId", "User", "id"),
          children: s.hasMany("id", "User", "parentId"),
        }),
      }),
      to: s.database({
        User: s.model({
          id: s.integer().autoincrement(),
          parentId: s.integer().nullable(),
        }),
      }),
    });
    expect(src).toMatchInlineSnapshot(`
      "BEGIN EXCLUSIVE TRANSACTION;
      CREATE TABLE \\"transferusers\\" (
        \\"id\\" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
        \\"parentId\\" INTEGER
      );
      INSERT INTO transferusers ( \\"id\\", \\"parentId\\" )
        SELECT \\"id\\", \\"parentId\\"
        FROM \\"users\\";
      DROP TABLE users;
      ALTER TABLE transferusers RENAME TO users;
      COMMIT TRANSACTION;"
    `);
  });

  it("ignores a renamed relation", () => {
    const src = generateMigration({
      from: s.database({
        User: s.model({
          id: s.integer().autoincrement(),
          parentId: s.integer().nullable(),
          relation0: s.belongsTo("parentId", "User", "id"),
        }),
      }),
      to: s.database({
        User: s.model({
          id: s.integer().autoincrement(),
          parentId: s.integer().nullable(),
          parent: s.belongsTo("parentId", "User", "id"),
          children: s.hasMany("parentId", "User", "id"),
        }),
      }),
    });
    expect(src).toMatchInlineSnapshot(`""`);
  });
});
