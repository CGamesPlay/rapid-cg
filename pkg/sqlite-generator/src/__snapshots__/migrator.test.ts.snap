// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateMigration adds a column 1`] = `"ALTER TABLE \\"tbl\\" ADD COLUMN \\"extra\\" TEXT NOT NULL;"`;

exports[`generateMigration adds a table 1`] = `
"CREATE TABLE \\"tbl\\" (
  \\"id\\" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL
);"
`;

exports[`generateMigration changes a column 1`] = `
"BEGIN EXCLUSIVE TRANSACTION;
CREATE TABLE \\"transfertbl\\" (
  \\"col\\" TEXT NOT NULL,
  \\"extra\\" TEXT
);
INSERT INTO transfertbl ( \\"col\\", \\"extra\\" )
  SELECT \\"col\\", \\"extra\\"
  FROM \\"tbl\\";
DROP TABLE tbl;
ALTER TABLE transfertbl RENAME TO tbl;
COMMIT TRANSACTION;"
`;

exports[`generateMigration drops a column 1`] = `"ALTER TABLE \\"tbl\\" DROP COLUMN \\"extra\\";"`;

exports[`generateMigration drops a table 1`] = `"DROP TABLE \\"tbl\\";"`;

exports[`generateMigration supports all column types 1`] = `
"CREATE TABLE \\"tbl\\" (
  \\"text\\" TEXT NOT NULL,
  \\"integer\\" INTEGER NOT NULL,
  \\"date\\" TEXT NOT NULL,
  \\"uuid\\" TEXT NOT NULL,
  \\"json\\" TEXT NOT NULL
);"
`;
